### PR TITLE
fix(datasets): make connection_args optional

### PR DESCRIPTION
**Pull request by jerome-asselin-buspatrol**
## Description
New `connection_args` argument did not have a default value. Hence existing applications fail.

## Development notes
Set the default to `None`.

## Checklist

- [No] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [No] Updated the documentation to reflect the code changes
- [Yes] Added a description of this change in the relevant `RELEASE.md` file
- [No] Added tests to cover my changes

### CODE REVIEW

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py [hunk 1]
```diff
@@ -225,11 +225,12 @@ def __init__(  # noqa: PLR0913
...
     @classmethod
-    def create_connection(cls, connection_str: str, connection_args: dict) -> None:
+    def create_connection(cls, connection_str: str, connection_args: dict = None) -> None:
         """Given a connection string, create singleton connection
         to be used across all instances of ``SQLTableDataset`` that
         need to connect to the same source.
         """
+        connection_args = connection_args or {}
         try:
             engine = create_engine(connection_str, **connection_args)
```
Comment (jerome-asselin-buspatrol) [inline kedro-datasets/kedro_datasets/pandas/sql_dataset.py:235]: Note: I do question how/if this `connection_args` should relate to `self._connection_args`, which is defined and used in other locations in that file.

Comment (merelcht) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py:235]: Yes it does. `create_connection` gets called within the `engine` property and that's where you can see that for `connection_args` it passes `self._connection_args`: https://github.com/kedro-org/kedro-plugins/blob/main/kedro-datasets/kedro_datasets/pandas/sql_dataset.py#L248

**Comment by merelcht** [comment]
Thanks for this contribution @jerome-asselin-buspatrol. I think this generally looks fine, but have asked @noklam and @deepyaman to review as well as they've touched this dataset before.

We require all commits to be signed-off to make the DCO check pass, you can find instructions here: https://github.com/kedro-org/kedro-plugins/pull/586/checks?check_run_id=22124623329 on how to fix this.

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py (hunk 1 above)
Comment (jerome-asselin-buspatrol) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py:235]: Ok. Therefore, what is the purpose of having added a `connection_args` argument in `create_connection()` in the first place?

Shouldn't we simply have called
```
engine = create_engine(connection_str, **self._connection_args)
```
and left `create_connection()` with a single argument (`connection_str`)?

**Comment by jerome-asselin-buspatrol** [comment]
My apologies, due to other commitments, I will not be able to address the github action issue in the next week. Feel free to pick it up if needed earlier, otherwise I’ll handle it in about a week.

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py (hunk 1 above)
Comment (noklam) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py:235]: With the same argument, connection_str can be replaced with self._connection_str.

I think the main reason for this is connection are shared in `cls.engines`, but not an object attribute. The class property `engine` is using `self._connection_str` and `self._connection_args` already.

Comment (jerome-asselin-buspatrol) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py:238]: Would you like me to apply required changes to remove both `connection_str` and `connection_args` arguments from `create_connection()`?

**Comment by jerome-asselin-buspatrol** [comment]
I'm not sure why the linters have "fixed" so many init files: that had nothing to do with my work.

Also, I find myself challenged to resolve the conflicting files. There's something I don't understand.
image

**Comment by noklam** [comment]
@jerome-asselin-buspatrol Can you try to base off the latest `main` and make changes to this PR? (or create a new one if that's easier for you). I think something is wrong here as you can see there are 300+ lines of change and touching things like `kedro-telemetry` readme, maybe you rebased from some old `main` or so.

**Comment by astrojuanlu** [comment]
CI passing, ready for review

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py [hunk 2]
```diff
@@ -498,11 +501,14 @@ def __init__(  # noqa: PLR0913
             self.adapt_mssql_date_params()
 
     @classmethod
-    def create_connection(cls, connection_str: str, connection_args: dict) -> None:
+    def create_connection(
+        cls, connection_str: str, connection_args: dict = None
```
Comment (noklam) [inline kedro-datasets/kedro_datasets/pandas/sql_dataset.py:505]:
```suggestion
        cls, connection_str: str, connection_args: Optional[dict] = None
```

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py [hunk 3]
```diff
@@ -226,11 +226,14 @@ def __init__(  # noqa: PLR0913
         self.metadata = metadata
 
     @classmethod
-    def create_connection(cls, connection_str: str, connection_args: dict) -> None:
+    def create_connection(
+        cls, connection_str: str, connection_args: dict = None
```
Comment (noklam) [inline kedro-datasets/kedro_datasets/pandas/sql_dataset.py:230]:
```suggestion
        cls, connection_str: str, connection_args: Optional[dict] = None
```

**Review by noklam — COMMENTED** [review]
Almost ready, minor fix about type hint and ready to go.

File: kedro-datasets/kedro_datasets/pandas/sql_dataset.py (hunk 2 above)
Comment (jerome-asselin-buspatrol) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py]: Note that the linter has automatically replaced the `Optional` use.

Comment (noklam) [reply kedro-datasets/kedro_datasets/pandas/sql_dataset.py]: that's fine, we actually preferred this I forgot this syntax is available already.

**Review by noklam — APPROVED** [review]
Thanks for making the PR !

**Comment by jerome-asselin-buspatrol** [comment]
> Thanks for making the PR !

My pleasure! And my apologies for all the back and forth on this rather trivial update. 🙏

**Review by merelcht — APPROVED** [review]
Thanks for the contribution @jerome-asselin-buspatrol ⭐

**Pull request merged**